File: 백준/Gold/2473. 세 용액/세 용액.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';

// const input = fs.readFileSync(0, 'utf-8').toString().trim().split('\n');
const input = fs.readFileSync(filePath).toString().trim().split('\n');

const arr = input[1].split(' ').map(e => +e).sort((a, b) => a - b);

let criteria = Infinity, ans;
for (let i = 0; i < arr.length; i++) {
    find(arr[i], i + 1);
}
console.log(ans.sort((a, b) => a - b).join(' '));

function find(fixed, start) {
    let end = arr.length - 1;

    let currentSum;

    while (start < end) {
        currentSum = fixed + arr[start] + arr[end];
        if (criteria > Math.abs(currentSum)) {
            criteria = Math.abs(currentSum);
            ans = [fixed, arr[start], arr[end]]
        }
        if (currentSum < 0) {
            start += 1
        } else if (currentSum === 0) {
            break;
        } else {
            end -= 1;
        }
    }
}

File: 백준/Gold/2473. 세 용액/세 용액.py
import sys

# print = sys.stdout.write
input = sys.stdin.readline

N = int(input().rstrip())

numbers = sorted(map(int, input().split()))


def main():
    ans = []
    gap = sys.maxsize
    for i in range(N):
        left, right = i + 1, N - 1

        while left < right:
            current_sum = numbers[i] + numbers[left] + numbers[right]
            if abs(current_sum) < abs(gap):
                ans = [numbers[i], numbers[left], numbers[right]]
                gap = abs(current_sum)

            if current_sum < 0:
                left += 1
            elif current_sum > 0:
                right -= 1
            else:
                print(*ans)
                return

    print(*ans)


if __name__ == "__main__":
    main()
